Coalesce anchoring status updates

robertknight opened the pull request:
Profiling the test case in #556 showed significant overhead from the
processing involved in or triggered by each `UPDATE_ANCHOR_STATUS`
action handled by the store. Previously one `UPDATE_ANCHOR_STATUS`
action was dispatched for each annotation whose anchoring status
changed.

Improve the situation by making the `UPDATE_ANCHOR_STATUS` action handle
updates for multiple annotations at once and modifying the `frameSync`
service to coalesce anchoring status updates from the host page into
a smaller number of `UPDATE_ANCHOR_STATUS` actions.

This commit also takes the opportunity to remove the logic to update
annotation tags when handling this action since it is no longer
required.

With this change the test case in #556 is not perfectly smooth, but it is significantly better and good enough to be usable on many systems.

Fixes #556

seanh commented on src/sidebar/frame-sync.js:
```diff
@@ -126,12 +128,24 @@ function FrameSync($rootScope, $window, Discovery, annotationUI, bridge) {
+      annotationUI.updateAnchorStatus(anchoringStatusUpdates);
+      $rootScope.$broadcast(events.ANNOTATIONS_SYNCED, Object.keys(anchoringStatusUpdates));
+      anchoringStatusUpdates = {};
+    }, 10);
```
Ok, so the anonymous function passed to `debounce()` will be called at most once every 10ms, no matter how many times we try to call it. And when it does finally get called, it uses `anchoringStatusUpdates` to update `annotationUI` and broadcast `ANNOTATIONS_SYNCED`, and then it resets `anchoringStatusUpdates` to empty.

robertknight replied:
Yes, exactly.

seanh commented on src/sidebar/frame-sync.js:
```diff
@@ -126,12 +128,24 @@ function FrameSync($rootScope, $window, Discovery, annotationUI, bridge) {
-        annotationUI.updateAnchorStatus(null, event.tag, event.msg.$orphan);
-        $rootScope.$broadcast(events.ANNOTATIONS_SYNCED, [event.tag]);
+        anchoringStatusUpdates[event.tag] = event.msg.$orphan ? 'orphan' : 'anchored';
+        scheduleAnchoringStatusUpdate();
```
Add this is where we add things in to `anchoringStatusUpdates` for the debounced function `scheduleAnchoringStatusUpdate()` to pick them up. And here we freely call `scheduleAnchoringStatusUpdate()` as often as we want, safe in the knowledge the it'll only actually be called at most once every 10ms.

robertknight replied:
Yes, indeed.

seanh commented on src/sidebar/reducers/annotations.js:
```diff
@@ -173,17 +173,16 @@ var update = {
-          $tag: action.tag,
-        });
-      } else {
+      if (!action.statusUpdates.hasOwnProperty(annot.$tag)) {
```
Ok, so here we're looping over _all_ of the annotations in the sidebar (I guess this is why it's slow, to call this function too often) and for each annotation, _if_ the annotation is in the status updates, we proceed

robertknight replied:
> Ok, so here we're looping over all of the annotations in the sidebar (I guess this is why it's slow, to call this function too often)

The current data structure that stores annotations in the sidebar is a list. This is indeed certainly suboptimal for certain operations. In practice however, it wasn't the (main) issue here from the profiling I looked at.

Rather the problem is that for every action that goes through the store, there are a bunch of processing steps that get triggered. These include Angular digest cycles (though these are debounced using `$applyAsync`) to trigger UI updates, plus any subscribers to state changes. Collectively this all adds up. It can be optimized, but batching updates is always going to be more efficient.

seanh commented on src/sidebar/reducers/annotations.js:
```diff
@@ -173,17 +173,16 @@ var update = {
+        return Object.assign({}, annot, { $anchorTimeout: true });
+      } else {
+        return Object.assign({}, annot, { $orphan: state === 'orphan' });
+      }
```
And for each annotation that gets updated, we _either_ set `$anchorTimeout: true` _or_ we set `$orphan` to either `true` or `false`. Here I'm confused. Isn't "anchor timeout" and "orphan" kind of the same thing? And I'm not sure why this PR is changing this block of code, from the `if (match)` block that was here before.

robertknight replied:
Other parts of the code make use of `$anchorTimeout`. I think we might indeed be able to have _just_ an `$orphan` flag in future, but to keep this PR smaller I preserved the existing behaviour of this action.

seanh replied:
The `if (state === 'timeout')` confuses me a little. This wasn't part of the logic of the previous code. It seems that it previously set _both_ `$anchorTimeout` and `$orphan`, but now it sets _either_ one or the other, depending on this new if?

robertknight replied:
Actually this was a bug in the previous code, when the `UPDATE_ANCHOR_STATUS` action was dispatched from this file, the `isOrphan` property was missing from the action, so `$orphan` was always being set to `undefined` here. That happens to be a no-op, but conceptually it is a bug.

seanh commented on src/sidebar/reducers/annotations.js:
```diff
@@ -265,19 +264,16 @@ function addAnnotations(annotations, now) {
+            statusUpdates[orphan.$tag] = 'timeout';
+            return statusUpdates;
+          }, {});
+        dispatch(updateAnchorStatus(statusUpdates));
```
I have no idea what is happening here

robertknight replied:
Thanks for the feedback. I rewrote this piece of code. Please let me know if it seems clearer now.

robertknight commented on src/sidebar/reducers/annotations.js:
```diff
@@ -262,22 +261,21 @@ function addAnnotations(annotations, now) {
+          updates[ann.$tag] = 'timeout';
+          return updates;
+        }, {});
+        dispatch(updateAnchorStatus(anchorStatusUpdates));
```
This is responsible for marking anchoring as having timed-out for newly loaded annotations if they fail to anchor within a certain time period. I modified this initially to dispatch _one_ action with N updates instead of N actions with one update each. In a second commit I rewrote this whole section for clarity in response to CR feedback.

seanh replied:
I see, so this is a function that runs after 500ms and orphans any annotations that haven't anchored yet. It'll end up calling the `UPDATE_ANCHOR_STATUS` function above, which will set `$anchorTimeout: true`. It seems odd that this is in the sidebar code rather than in the annotator code, but I appreciate that you've rewritten it in-place for clarity.

robertknight replied:
The reason it is in the sidebar rather than the annotator code is because the "annotator" code runs in the context of the host page and there are various ways in which that could cause the timeout code to either not run or to fail to execute within the timeout period. For example, a news website bogged down by a ton of crappy JS and ads.

Putting the code in the sidebar avoids some of these failure scenarios and, depending on the browser's JS execution scheduling, avoids its execution being blocked by code in the host page.

robertknight commented on src/sidebar/reducers/annotations.js:
```diff
@@ -173,17 +173,16 @@ var update = {
 
   UPDATE_ANCHOR_STATUS: function (state, action) {
     var annotations = state.annotations.map(function (annot) {
-      var match = (annot.id && annot.id === action.id) ||
```
It used to be the case that some annotations (existing ones that had not yet been anchored) had only IDs, some had only tags (newly created but not yet saved ones) and some had both (saved annotations that had been anchored).

Every annotation now has a `$tag` as soon as it is added to the store, so we no longer need to check for either an ID or a tag.

seanh replied:
Ah, I see: "This commit also takes the opportunity to remove the logic to update annotation tags when handling this action since it is no longer required." This caused me a fair bit of confusion (since the change seemed to have nothing to do with coalescing updates), if this could have been done as a separate PR or at least a separate commit it might have been easier to follow.

robertknight commented on src/sidebar/test/frame-sync-test.js:
```diff
@@ -189,16 +189,39 @@ describe('FrameSync', function () {
+      assert.calledWith(fakeAnnotationUI.updateAnchorStatus, { t1: 'anchored' });
+    });
+
+    it('coalesces multiple "sync" messages', () => {
```
This is a test to check that multiple sync messages in quick succession result in a single action being dispatched.

robertknight commented on src/sidebar/test/frame-sync-test.js:
```diff
@@ -189,16 +189,39 @@ describe('FrameSync', function () {
       $rootScope.$on(events.ANNOTATIONS_SYNCED, onSync);
 
       fakeBridge.emit('sync', [{tag: 't1', msg: {$orphan: false}}]);
+      clock.tick(20);
```
This is needed because emitting the `ANNOTATIONS_SYNCED` event no longer happens synchronously, due to debouncing, when a "sync" message is received by the sidebar.

robertknight commented on src/sidebar/test/frame-sync-test.js:
```diff
@@ -189,16 +189,39 @@ describe('FrameSync', function () {
   });
 
   context('when anchoring completes', function () {
+    var clock = sinon.stub();
```
Due to the introduction of debouncing, we now need to play god with time in the tests.

robertknight commented on src/sidebar/test/annotation-ui-test.js:
```diff
@@ -483,28 +491,12 @@ describe('annotationUI', function () {
   });
 
   describe('#updatingAnchorStatus', function () {
-    it("updates the annotation's tag", function () {
```
All annotations now have tags as soon as they are added to the store, so `updateAnchorStatus` no longer does what this test checks.

robertknight commented on src/sidebar/test/annotation-ui-test.js:
```diff
@@ -483,28 +491,12 @@ describe('annotationUI', function () {
       assert.equal(annotationUI.getState().annotations[0].$orphan, true);
     });
-
-    it('updates annotations by tag', function () {
```
This test is now irrelevant since there is only one way to reference which annotation was anchored / failed to anchor, and that is covered by the test above.

robertknight commented:
Thanks for the review @seanh. I updated a piece of code that you flagged up as unclear and added some comments to a few other things. Please take a look.
